src/NovelAudioService.cpp

```diff
@@ -0,0 +1,367 @@
+  }
+}
+
+void NovelAudioService::setSoundDistance(std::string soundName, int distance) {
+    if (_channelMap.find(soundName) != _channelMap.end() && _channelMap[soundName] != -1)
+  {
+    Mix_SetDistance(_channelMap[soundName], distance);
+  }
+  else
+  {
+    std::cerr << soundName << " is not allocated to a channel already!" << std::endl;
+  }
+}
+
+void NovelAudioService::setSoundPanning(std::string soundName, int leftChannelVolume, int rightChannelVolume) {
+    if (_channelMap.find(soundName) != _channelMap.end() && _channelMap[soundName] != -1)
+  {
+    Mix_SetPanning(_channelMap[soundName], leftChannelVolume, rightChannelVolume);
+  }
+  else
+  {
+    std::cerr << soundName << " is not allocated to a channel already!" << std::endl;
+  }
+}
+
+void NovelAudioService::loadMusic(std::string input) {
+  auto exists = _music.find(input);
+  if (exists != _music.end()) return;
+
+  Mix_Music* newMusic = Mix_LoadMUS(input.c_str());
+  if (newMusic != nullptr)
+  {
+    _music[input] = newMusic;
+  }
+  else
+  {
+    std::cerr << Mix_GetError() << std::endl;
+  }
+}
+
+void NovelAudioService::unloadMusic(std::string input) {
+  auto existingMusic = _music.find(input);
+  if (existingMusic == _music.end()) return;
+
+  Mix_FreeMusic(_music[input]);
+  _music.erase(existingMusic);
+}
+
+void NovelAudioService::playMusic() {
+  if (Mix_PausedMusic()) Mix_ResumeMusic();
+}
+
+void NovelAudioService::playMusic(std::string musicName) {
+  auto existingMusic = _music.find(musicName);
+  if (existingMusic == _music.end())
+  {
+    loadMusic(musicName);
+    existingMusic = _music.find(musicName);
+    if (existingMusic == _music.end()) return;
+  }
+  _channelMap[musicName] = -3;
+  Mix_PlayMusic(_music[musicName],-1);
+  _musicTime = SDL_GetPerformanceCounter();
+}
+
+void NovelAudioService::playMusic(std::string musicName, int loops) {
+
+  auto existingMusic = _music.find(musicName);
+  if (existingMusic == _music.end())
+  {
+    loadMusic(musicName);
+    existingMusic = _music.find(musicName);
+    if (existingMusic == _music.end()) return;
+  }
+  _channelMap[musicName] = -3;
+  Mix_PlayMusic(_music[musicName], loops-1);
+}
+
+void NovelAudioService::pauseMusic() {
+  if (!Mix_PausedMusic())
+  {
+    Mix_PauseMusic();
+    _musicPausedTime = SDL_GetPerformanceCounter();
+  }
+}
+
+void NovelAudioService::stopMusic() {
+  Mix_HaltMusic();
+  _musicPausedTime = SDL_GetPerformanceCounter();
+}
+
+void NovelAudioService::setMusicVolume(float value) {
+  Mix_VolumeMusic(convertToMixVolume(value));
+}
+
+void NovelAudioService::fadeMusicInOnce(std::string musicName, int ms) {
+  if (!Mix_PlayingMusic())
+  {
+    Mix_FadeInMusic(_music[musicName], 0, ms);
+  }
+  else
+  {
+    Mix_FadeOutMusic(1000);
+    switch(Mix_GetMusicType(_music[musicName]))
+    {
+      case MUS_OGG:
+      {
+        double pos = (double)((_musicPausedTime - _musicTime)*1000) / (double)SDL_GetPerformanceFrequency();
```

Comment:
Here is another reason why having a `Timestamp` type would be good.

There is a subtle, but non obvious bug here that deals with floating-point values and timestamps in general. Namely around `Frequency` being significantly larger than the `TicksPerSecond`, which you've defined to be `1000`. For example, on my box, `Frequency` is `10'000'000`, which means that this has a risk of producing a really tiny result and incurring rounding error.

Reply:
Done, issue created!